### ISSUE TITLE
Hyperparameter Optimization

### ISSUE THREAD

**Issue by pravanjang** [pravanjang/stock-prediction-system#18]
**Code to Generate:**
Complete script `models/optimize_hyperparams.py`:

**Required Functions:**
```python
def setup_optuna_study()
def objective_function(trial):
    """
    Optimize:
    - BGRU hidden dimensions: [64, 128, 256]
    - Number of BGRU layers: [1, 2, 3]
    - Dropout rates: [0.2, 0.3, 0.4, 0.5]
    - Learning rate: [1e-4, 5e-4, 1e-3]
    - Batch size: [32, 64, 128]
    """
    
def train_with_best_params(best_params)
def apply_class_balancing(train_df)
```

**Optimization Strategy:**
- Use Optuna for Bayesian optimization
- 30 trials maximum
- Optimize for validation accuracy
- Save best hyperparameters to `models/best_hyperparams.json`

**Class Balancing:**
- Check class distribution
- If imbalanced (>55:45), apply weighted loss
- Calculate class weights: `weight = total / (n_classes * count_per_class)`

**Execution:**
```bash
python models/optimize_hyperparams.py --data data/processed/train_final.csv --n_trials 30
```

**Output:**
- `models/best_hyperparams.json`
- `models/optimization_history.csv`
- `models/checkpoints/optimized_bgru.pt`
